### PR TITLE
Staff f20 add archived course repo

### PR DESCRIPTION
**Pull request by pconrad**
In this PR we add support for an archived course repository, a MongoDB database of all courses for past quarters.

The MongoDB database should be used (instead of the API) for searches that span multiple quarters; it will be faster, and it supports more kinds of queries.

We add a History menu where searches based on the historical data can be placed.  From this menu, we add a basic search that is identical to the main search except that it doesn't include level.

### CODE REVIEW

File: src/test/java/edu/ucsb/courses/repositories/RunWith.java [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package edu.ucsb.courses.repositories;
+
+public @interface RunWith {
+
+}
```
Comment (scottpchow23) [inline src/test/java/edu/ucsb/courses/repositories/RunWith.java:5]: Did we still want this file? I'm not sure what this is doing.

File: src/main/java/edu/ucsb/courses/documents/Section.java [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+        }
+        Section section = (Section) o;
+        return Objects.equals(enrollCode, section.enrollCode) && Objects.equals(section, section.section) && Objects.equals(session, section.session) && Objects.equals(classClosed, section.classClosed) && Objects.equals(courseCancelled, section.courseCancelled) && Objects.equals(gradingOptionCode, section.gradingOptionCode) && Objects.equals(enrolledTotal, section.enrolledTotal) && Objects.equals(maxEnroll, section.maxEnroll) && Objects.equals(secondaryStatus, section.secondaryStatus) && departmentApprovalRequired == section.departmentApprovalRequired && instructorApprovalRequired == section.instructorApprovalRequired && Objects.equals(restrictionLevel, section.restrictionLevel) && Objects.equals(restrictionMajor, section.restrictionMajor) && Objects.equals(restrictionMajorPass, section.restrictionMajorPass) && Objects.equals(restrictionMinor, section.restrictionMinor) && Objects.equals(restrictionMinorPass, section.restrictionMinorPass) && Objects.equals(concurrentCourses, section.concurrentCourses) && Objects.equals(timeLocations, section.timeLocations) && Objects.equals(instructors, section.instructors);
```
Comment (scottpchow23) [inline src/main/java/edu/ucsb/courses/documents/Section.java:358]: We should ideally use `EqualsBuilder` to reduce the number of tests we need to do.

File: src/test/java/edu/ucsb/courses/documents/SectionTests.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    }
+
+    @Test
+    public void test_getAndSetConcurrentCourses() {
+        Section s = new Section();
+        List<String> concurrentCourses = new ArrayList<String>();
+
+        s.setConcurrentCourses(concurrentCourses);
+        assertEquals(concurrentCourses, s.getConcurrentCourses());
+    }
```
Comment (scottpchow23) [inline src/test/java/edu/ucsb/courses/documents/SectionTests.java:30]: Looks like we need to finish adding tests for `timeLocations` and `instructors`.

**Review by scottpchow23 — CHANGES_REQUESTED** [review]
I pointed out the few missing pieces but if we make those changes we can definitely get this in.

File: src/main/java/edu/ucsb/courses/documents/Section.java (hunk 2 above)
Comment (pconrad) [reply src/main/java/edu/ucsb/courses/documents/Section.java:358]: Yes... I'm moving to that one repo at a time.

File: src/test/java/edu/ucsb/courses/documents/SectionTests.java (hunk 3 above)
Comment (pconrad) [reply src/test/java/edu/ucsb/courses/documents/SectionTests.java:30]: Yes... that's also part of the WIP

File: src/test/java/edu/ucsb/courses/repositories/RunWith.java (hunk 1 above)
Comment (pconrad) [reply src/test/java/edu/ucsb/courses/repositories/RunWith.java]: Deleted

File: javascript/src/main/components/BasicCourseSearch/CourseSearchFormQtrDeptOnly.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+import { Form, Button } from "react-bootstrap";
+
+const CourseSearchFormQtrDeptOnly = ({ setCourseJSON, fetchJSON }) => {
+
+    const [quarter, setQuarter] = useState("20211");
+    const [department, setDepartment] = useState("CMPSC");
+
+    const handleSubmit = (event) => {
+        event.preventDefault();
+        console.log("submit pressed");
```
Comment (scottpchow23) [inline javascript/src/main/components/BasicCourseSearch/CourseSearchFormQtrDeptOnly.js:11]: I wonder if we still need this?

File: javascript/src/main/services/courseSearches.js [hunk 5]
```diff
@@ -8,5 +8,13 @@ import fetch from "isomorphic-unfetch";
     return courseJSON;
 };
 
-export { fetchBasicCourseJSON };
+const fetchBasicCourseHistoryJSON = async (event, fields) => {
+    const url = `/api/public/history/basicsearch?qtr=${fields.quarter}&dept=${fields.department}`;
+    console.log(`fetching JSON, url=${url}`);
```
Comment (scottpchow23) [inline javascript/src/main/services/courseSearches.js:13]: I wonder if we still need this?

File: javascript/src/main/services/courseSearches.js [hunk 6]
```diff
@@ -8,5 +8,13 @@ import fetch from "isomorphic-unfetch";
...
+    const courseJSON = (await (await fetch(url)).json())
+    console.log(`fetch returned, courseJSON=${courseJSON}`);
```
Comment (scottpchow23) [inline javascript/src/main/services/courseSearches.js:15]: I wonder if we still need this?

File: src/main/java/edu/ucsb/courses/controllers/HistorySearchController.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+@RestController
+@RequestMapping("/api/public/history")
+public class HistorySearchController {
+    private final Logger logger = LoggerFactory.getLogger(HistorySearchController.class);
+
+    private ObjectMapper mapper = new ObjectMapper();
+
+    // @Autowired
+    // UCSBCurriculumService ucsbCurriculumService;
```
Comment (scottpchow23) [inline src/main/java/edu/ucsb/courses/controllers/HistorySearchController.java:34]: I wonder if we still need this?

File: src/test/java/edu/ucsb/courses/controllers/HistorySearchControllerTests.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;
+
+import java.util.ArrayList;
+import java.util.List;
+
+// @Import(SecurityConfig.class) applies the security rules 
+// so that /api/public/** endpoints don't require authentication.
+// Otherwise you may get authorization errors when running the test
+
```
Comment (scottpchow23) [inline src/test/java/edu/ucsb/courses/controllers/HistorySearchControllerTests.java:30]: Oh, is this an alternative to using `@WithMockUser`? If so, I love this.

**Review by scottpchow23 — COMMENTED** [review]
Couple of nitpicks for some `console.log`s and some commented code but otherwise this looks awesome!

File: javascript/src/main/components/BasicCourseSearch/CourseSearchFormQtrDeptOnly.js (hunk 4 above)
Comment (pconrad) [reply javascript/src/main/components/BasicCourseSearch/CourseSearchFormQtrDeptOnly.js:11]: Probably not.   Removing it...

File: javascript/src/main/services/courseSearches.js (hunk 5 above)
Comment (pconrad) [reply javascript/src/main/services/courseSearches.js:13]: console.log's removed

File: src/test/java/edu/ucsb/courses/controllers/HistorySearchControllerTests.java (hunk 8 above)
Comment (pconrad) [reply src/test/java/edu/ucsb/courses/controllers/HistorySearchControllerTests.java:30]: Possibly....    depending on the circumstances.

**Review by scottpchow23 — APPROVED** [review]

**Review by tkomarlu — DISMISSED** [review]
LGTM

File: javascript/src/main/services/courseSearches.js [hunk 9]
```diff
@@ -2,11 +2,15 @@ import fetch from "isomorphic-unfetch";
 
  const fetchBasicCourseJSON = async (event, fields) => {
     const url = `/api/public/basicsearch?qtr=${fields.quarter}&dept=${fields.department}&level=${fields.level}`;
-    console.log(`fetching JSON, url=${url}`);
     const courseJSON = (await (await fetch(url)).json())
-    console.log(`fetch returned, courseJSON=${courseJSON}`);
     return courseJSON;
```
Comment (btk5h) [inline javascript/src/main/services/courseSearches.js:6]: Same concern:
```suggestion
    const courseResponse = await fetch(url);
    return courseResponse.json();
```

File: javascript/src/main/services/courseSearches.js (hunk 6 above)
Comment (pconrad) [reply javascript/src/main/services/courseSearches.js]: console.log's removed

Comment (pconrad) [reply javascript/src/main/services/courseSearches.js]: Removed

File: javascript/src/main/services/courseSearches.js (hunk 9 above)
Comment (pconrad) [reply javascript/src/main/services/courseSearches.js]: fixed

**Comment by pconrad** [comment]
Was reviewed by @btk5h @scottpchow23 and @tkomarlu ... it wasn't "re-reviewed" after I fixed the items that @btk5h and @scottpchow23  mentioned, but they were relatively minor, and all the tests are passing.

**Pull request merged**